xen/arch/arm/coproc/plat/gx6xxx.c

```diff
@@ -0,0 +1,303 @@
+#define RGXFW_CR_IRQ_STATUS_EVENT_EN                  RGX_CR_MIPS_WRAPPER_IRQ_STATUS_EVENT_EN
+#define RGXFW_CR_IRQ_CLEAR                            RGX_CR_MIPS_WRAPPER_IRQ_CLEAR
+#define RGXFW_CR_IRQ_CLEAR_MASK                       RGX_CR_MIPS_WRAPPER_IRQ_CLEAR_EVENT_EN
+#else
+#define RGX_CR_META_SP_MSLVIRQSTATUS                  (0x0AC8U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_MASKFULL         (IMG_UINT64_C(0x000000000000000C))
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT3_SHIFT  (3U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT3_CLRMSK (0XFFFFFFF7U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT3_EN     (0X00000008U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_SHIFT  (2U)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_CLRMSK (0XFFFFFFFBU)
+#define RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_EN     (0X00000004U)
+
+#define RGXFW_CR_IRQ_STATUS                           RGX_CR_META_SP_MSLVIRQSTATUS
+#define RGXFW_CR_IRQ_STATUS_EVENT_EN                  RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_EN
+#define RGXFW_CR_IRQ_CLEAR                            RGX_CR_META_SP_MSLVIRQSTATUS
+#define RGXFW_CR_IRQ_CLEAR_MASK                       RGX_CR_META_SP_MSLVIRQSTATUS_TRIGVECT2_CLRMSK
+#endif
+
+static int gx6xxx_read(struct vcpu *v, mmio_info_t *info,
+                       register_t *r, void *priv)
+{
+    struct mmio *mmio = priv;
+    struct vcoproc_rw_context ctx;
+    static int start = 1;
+    uint32_t *vaddr;
+    unsigned long flags;
+
+    vcoproc_get_rw_context(v->domain, mmio, info, &ctx);
+    spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    if (ctx.offset == RGXFW_CR_IRQ_STATUS) {
+        struct vgx6xxx_info *vinfo = (struct vgx6xxx_info *)ctx.vcoproc->priv;
+
+        *r = vinfo->irq_status;
+        spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
+        return 1;
+    }
+    if ( unlikely(start) ) {
+        /* TODO: find condition to start the scheduler */
+        start = 0;
+        spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
+        vcoproc_sheduler_vcoproc_wake(ctx.coproc->sched, ctx.vcoproc);
+        spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    }
+    vaddr = (uint32_t *)((char *)mmio->base + ctx.offset);
+    *r = readl(vaddr);
+    spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
+    return 1;
+}
+
+static int gx6xxx_write(struct vcpu *v, mmio_info_t *info,
+                        register_t r, void *priv)
+{
+    struct mmio *mmio = priv;
+    struct vcoproc_rw_context ctx;
+    uint32_t *vaddr;
+    unsigned long flags;
+
+    vcoproc_get_rw_context(v->domain, mmio, info, &ctx);
+    spin_lock_irqsave(&ctx.coproc->vcoprocs_lock, flags);
+    if (ctx.offset == RGXFW_CR_IRQ_STATUS) {
+        struct vgx6xxx_info *vinfo = (struct vgx6xxx_info *)ctx.vcoproc->priv;
+
+        vinfo->irq_status = r;
+        spin_unlock_irqrestore(&ctx.coproc->vcoprocs_lock, flags);
+        return 1;
+    }
+    vaddr = (uint32_t *)((char *)mmio->base + ctx.offset);
```

Comment:
Do we need vaddr/cast here? Can't we write in a such way:
writel(r, mmio->base + ctx.offset); ?